[Sanjevi] iP

sanjevi13 opened the pull request:
# Duke Bot
>“The key is not to prioritize what’s on your schedule, but to schedule your priorities.” – Stephen Covey ([source](https://www.tameday.com/motivational-quotes-for-work-productivity/))

DukeBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. Download it from [here](https://github.com/sanjevi13/ip/releases/tag/A-Jar)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you:wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

sanjevi13 committed:
Implemented Level 1: Greet, Echo, Exit

sanjevi13 committed:
Implemented Level 2: Add, List

sanjevi13 committed:
Implemented Level 3: Mark as Done, fixed minor bugs from Level 2

sanjevi13 committed:
Implemented Level 4: ToDos, Events, Deadlines

sanjevi13 committed:
Implemented A-TextUiTesting: Automated Text UI Testing

sanjevi13 committed:
Implemented Level 5: Handle errors

sanjevi13 committed:
Implemented Level 6: Delete

sanjevi13 committed:
Added exception handling for Delete

sanjevi13 committed:
Implement Level 7: Save

sanjevi13 committed:
Implement Level 8: Dates and Times, improve Event class

sanjevi13 committed:
Merge branch 'branch-level-7'

sanjevi13 committed:
Merge branch 'branch-level-8' and resolve merge conflict

sanjevi13 committed:
Implement A-MoreOOP: Add Ui, Storage, Parser, Tasklist

sanjevi13 committed:
Implement A-Packages: Organize into Packages

sanjevi13 committed:
Merge remote-tracking branch 'origin/add-gradle-support' to master

sanjevi13 committed:
Implement A-Gradle: Use Gradle

sanjevi13 committed:
Implement A-JUnit: Add JUnit tests

sanjevi13 committed:
Fixed file path for duke.txt

sanjevi13 committed:
Implement A-JavaDoc: Add JavaDocs to most public methods

sanjevi13 committed:
Implement A-CodingStandard: Follow the Coding Standard

sanjevi13 committed:
Implement Level-9: Find

sanjevi13 committed:
Merge branch 'branch-A-CodingStandard' and resolve conflicts

sanjevi13 committed:
Merge branch 'branch-level-9' and resolve conflicts

4ndrelim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,99 @@
+package duke;
+import java.io.File;
```
I think the ordering of your import statements is consistent and easily understood (noticed this elsewhere too!) but perhaps you could consider adding a linespace for clear segregation so readers can easily tell the required imports from different packages at a glance 😄

4ndrelim commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,40 @@
+     * @return The string representation of the Deadline task that is stored in a text file.
+     */
+    @Override
+    public String sendOutputToFile() {
```
Just a minor naming convention - Semantically, the method isn't sending the string representation but rather, getting the string representation of the task to be subsequently stored in your file storage. Perhaps "getOutput" / "getRepresentation" would be more apt?

4ndrelim commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,99 @@
+            if (!file.exists()) {
+                file.getParentFile().mkdir();
+            }
+            if (file.exists()) {
```
I think there is slightly too many levels of nesting here and may affect readability. Perhaps you can restructure the conditionals here to have your `while` loop extracted out of the nesting? I feel that would also make the 'happy path' more prominent!

4ndrelim commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,295 @@
+     * @param input The input string.
+     * @throws DukeException if the input is invalid.
+     */
+    public void todoInputChecker(String input) throws DukeException {
```
I noticed some of your methods implementation in this file do not quite align with its semantic meaning, one might expect a boolean value to be returned by reading the javadocs. 
The name for the method as a noun may have circumvented this but it feels weird (using exceptions as a control flow) to have a method whose sole purpose is to throw errors. Perhaps an alternative approach would be to simply throw an error in the main execution if the user input evaluates to false for all your 'checkers'?

4ndrelim commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+/**
+ * Encapsulates the main Duke app.
+ */
+public class Duke {
```
Nice abstraction of each component!

4ndrelim reviewed (commented):
Overall, great job adhering to the java coding standard! Your code is neat and relatively easy to read 😄

Alvozyn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,60 @@
+    public void runApp() {
+        ui.showDuke();
+        ui.greetingMessage();
+        boolean enteredBye = false;
```
Line 38 naming convention: Boolean enterBye can be to look more like a question. (Eg. isByeEntered or hasEnteredBye)

Alvozyn commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,37 @@
+/**
+ * Encapsulates an Event as a specific type of Task.
+ */
+
```
Line 6: You can remove this blank line as javadocs should be directly above the class/method

Alvozyn commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    /**
+     * Parses the input string and modifies the list accordingly
```
Line 6: (Minor) End with fullstops

Alvozyn commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,153 @@
+package duke;
+import java.util.ArrayList;
```
Line2: There should be a new line between package and imports

Alvozyn commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,153 @@
+
+    /**
+     * Prints the tasks found using the query function.
+     * @param taskList The TaskList object containing the ArrayList of tasks
```
Line112: (Minor) Remember to end with fullstops

Alvozyn reviewed (commented):
Overall, your code is really easy to read and the javadocs is well documented! I learnt some stuff from you too that can help me write my javadocs better. :)

sanjevi13 committed:
Implement Level-10: GUI

sanjevi13 committed:
fix add deadline error

sanjevi13 committed:
Refactor into gui package and modify JavaDocs

Files in the gui package run the main application window, while files
in duke package handles user interaction (e.g. add tasks, mark tasks
as done/not done, delete tasks)

Refactoring the code improves readability and accessibility of code.

Javadocs needed to be modified as the previous commit saw the
modification of several return statements in UI and Parser.

sanjevi13 committed:
Implement A-Assertions: Use Assertions

sanjevi13 committed:
Implement A-CodeQuality: Improve Code Quality
Improve user interaction with Duke bot
When user is done with task list, the chatbot would close immediately
With the implementation of delay, user can now read the goodbye message.

sanjevi13 committed:
Merge pull request #2 from sanjevi13/branch-A-CodeQuality

Implement A-CodeQuality: Improve Code Quality

sanjevi13 committed:
Merge branch 'master' into branch-A-Assertions

sanjevi13 committed:
Merge pull request #3 from sanjevi13/branch-A-Assertions

Implement A-Assertions: Use Assertions

sanjevi13 committed:
Implement C-Help: Give help to users.
Provide in-App guidance to users.
Add a command to access help page.

New users may not know how to use the app. If they use 3 invalid
commands, a help message will be generated, and a command to access
the command list will be given.

With the addition of command list, accessibility will improve and new
users can learn how to use the app easily.

sanjevi13 committed:
Fix javadocs and modify some UI messages

sanjevi13 committed:
Improve code quality: Abstract some repeated methods as a single general
method to call multiple times.

sanjevi13 committed:
Fix delete function

sanjevi13 committed:
Add User Guide, Reformat event task back to old version
Reduce application exit wait time to 2s from 3s upon command

sanjevi13 committed:
Change Ui.png and link to download duke

sanjevi13 committed:
Update readme

sanjevi13 committed:
Update Readme final

sanjevi13 committed:
Update readme for submission

sanjevi13 committed:
Improve GUI: maximize/minimize window

sanjevi13 committed:
Fix minor bug for event addition
